xmtp_api_d14n/src/compat/d14n.rs

```diff
@@ -41,80 +48,315 @@ where
     E: std::error::Error + RetryableError + Send + Sync + 'static,
     P: Send + Sync + Client,
     C: Send + Sync + Client,
+    ApiError<E>: From<ApiError<<P as Client>::Error>>
+        + From<ApiError<<C as Client>::Error>>
+        + Send
+        + Sync
+        + 'static,
 {
     type Error = ApiError<E>;
+
     async fn upload_key_package(
         &self,
         request: UploadKeyPackageRequest,
     ) -> Result<(), Self::Error> {
-        todo!()
+        let envelope: ClientEnvelope = request.try_into()?;
+
+        PublishClientEnvelopes::builder()
+            .envelopes(vec![envelope])
+            .build()
+            .unwrap()
```

Comment:
ideally we don't unwrap at all, but can do this in a followup or create an issue for this